pkg/scale/scale.go

```diff
@@ -0,0 +1,230 @@
+	"github.com/appscode/stash/pkg/backup"
+	"github.com/appscode/stash/pkg/util"
+	apps "k8s.io/api/apps/v1beta1"
+	core "k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type Options struct {
+	Workload       api.LocalTypedReference
+	Namespace      string
+	Selector       string
+	DockerRegistry string // image registry for check job
+	ImageTag       string // image tag for check job
+}
+
+type Controller struct {
+	k8sClient kubernetes.Interface
+	opt       Options
+	locked    chan struct{}
+}
+
+var (
+	ZeroReplica = int32(0)
+	OneReplica  = int32(1)
+)
+
+func New(k8sClient kubernetes.Interface, opt Options) *Controller {
+	return &Controller{
+		k8sClient: k8sClient,
+		opt:       opt,
+	}
+}
+
+func (c *Controller) ScaleDownWorkload() error {
+
+	// scale down deployment to 0 replica
+	dpList, err := c.k8sClient.AppsV1beta1().Deployments(c.opt.Namespace).List(metav1.ListOptions{LabelSelector: c.opt.Selector})
+	if err == nil {
+		for _, dp := range dpList.Items {
+			_, _, err := apps_util.PatchDeployment(c.k8sClient, &dp, func(obj *apps.Deployment) *apps.Deployment {
+				if obj.Annotations == nil {
+					obj.Annotations = make(map[string]string)
+				}
+				obj.Annotations[util.AnnotationOldReplica] = strconv.Itoa(int(*dp.Spec.Replicas))
+				obj.Spec.Replicas = &ZeroReplica
+				return obj
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// scale down replication controller to 0 replica
+	rcList, err := c.k8sClient.CoreV1().ReplicationControllers(c.opt.Namespace).List(metav1.ListOptions{LabelSelector: c.opt.Selector})
+	if err == nil {
+		for _, rc := range rcList.Items {
+			_, _, err := core_util.PatchRC(c.k8sClient, &rc, func(obj *core.ReplicationController) *core.ReplicationController {
+				if obj.Annotations == nil {
+					obj.Annotations = make(map[string]string)
+				}
+				obj.Annotations[util.AnnotationOldReplica] = strconv.Itoa(int(*rc.Spec.Replicas))
+				obj.Spec.Replicas = &ZeroReplica
+				return obj
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// scale down replicaset to 0 replica
+	rsList, err := c.k8sClient.ExtensionsV1beta1().ReplicaSets(c.opt.Namespace).List(metav1.ListOptions{LabelSelector: c.opt.Selector})
+	if err == nil {
+		for _, rs := range rsList.Items {
+			_, _, err := ext_util.PatchReplicaSet(c.k8sClient, &rs, func(obj *extensions.ReplicaSet) *extensions.ReplicaSet {
+				if obj.Annotations == nil {
+					obj.Annotations = make(map[string]string)
+				}
+				obj.Annotations[util.AnnotationOldReplica] = strconv.Itoa(int(*rs.Spec.Replicas))
+				obj.Spec.Replicas = &ZeroReplica
+				return obj
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// delete all pods of daemonset and statefulset so that they restart with init container
+	podList, err := c.k8sClient.CoreV1().Pods(c.opt.Namespace).List(metav1.ListOptions{LabelSelector: c.opt.Selector})
```

Comment:
we are deleting all pods here, so what's the use of patch replica 0? can't we just patch to replica 1 for deploy, rs & rc and then delete all pods?

Reply:
This will remove only daemonset and statefulset's pod.  As we wanted to avoid manual deletion of pods, we avoided for dp,rs and rc by scaling down to 0.